### ISSUE TITLE
question about SArray.jl

### ISSUE THREAD

**Issue by githubtomtom** [JuliaArrays/StaticArrays.jl#721]
https://github.com/JuliaArrays/StaticArrays.jl/blob/3a9291df5407230ac8cce5aec025abaa068fd5d4/src/SArray.jl#L32

in above I see the use of generated function for outer constructors.

what is the difference (and disadvantage) using the following instead?
```
SArray{S, T, N}(x::Tuple) where {S<:Tuple, T, N} = SArray{S, T, N, tuple_prod(S) }(x)
```

also, in https://github.com/JuliaArrays/StaticArrays.jl/blob/3a9291df5407230ac8cce5aec025abaa068fd5d4/src/SDiagonal.jl#L11

why don't we define the following?
```
@inline SDiagonal{N,T}(a::Tuple) where {N,T} = Diagonal(SVector{N,T}(a))
```

thanks.

**Comment by githubtomtom** [comment]
also, in https://github.com/JuliaArrays/StaticArrays.jl/blob/3a9291df5407230ac8cce5aec025abaa068fd5d4/src/SDiagonal.jl#L11

why don't we define the following?
```
SDiagonal{N,T}(a::Tuple) where {N,T} = Diagonal(SVector{N,T}(a))
```

**Comment by c42f** [comment]
> what is the difference (and disadvantage) using the following instead?

The fact this is `@generated` is likely a historical artifact from a time when the `Base.@pure` annotation didn't exist (keep in mind that StaticArrays dates from julia 0.5 or so). You'd need to read the history to know for sure. At this stage we could very likely replace this with a normal function.

> why don't we define the following? `@inline SDiagonal [...]`

Again, likely historical (due to the syntax changes in the julia 0.6 -> 1.0 transition).

**Comment by githubtomtom** [comment]
thanks.

for the first case, maybe using generated function could push the computation of tuple_prod(S) into compilation time?

> On Jan 30, 2020, at 1:36 PM, Chris Foster <notifications@github.com> wrote:
> 
> what is the difference (and disadvantage) using the following instead?
> 
> The fact this is @generated is likely a historical artifact from a time when the Base.@pure annotation didn't exist (keep in mind that StaticArrays dates from julia 0.5 or so). You'd need to read the history to know for sure. At this stage we could very likely replace this with a normal function.
> 

but the second one is really mysterious to me.

**Comment by c42f** [comment]
> maybe using generated function could push the computation of tuple_prod(S) into compilation time

Indeed, that's exactly what that `$` does there: compute `tuple_prod` as part of the generator and splice the result into the AST.  But it's also very likely unnecessary because `tuple_prod` is now marked `@pure` (so the compiler will do the same thing without the `@generated` compile time overhead).

> the second one is really mysterious to me

The syntax for constructors of parametric types was very different in julia 0.5. If you read the old manuals for julia 0.6 vs 1.0 you may understand.

As I said: it's likely purely historical. I think your proposed rewording would be equivalent but clearer in the new syntax.
